Object caching `WP_Theme_JSON::compute_style_properties`

kt-12 opened the pull request:
Trac ticket: [59595](https://core.trac.wordpress.org/ticket/59595)

Incorporating feedback from - https://github.com/WordPress/wordpress-develop/pull/5567, I did the first version https://github.com/WordPress/wordpress-develop/pull/5567


## `theme_json` based caching based on initial feeback ( [commit](https://github.com/WordPress/wordpress-develop/pull/6392/commits/b7ba582722c08889064da8f1df6fc5f6a17dc9ef) )
* The commit included using, `theme_json` for caching. However `theme_json` being non-persistent had no benefit.  Instead, the cache key generation added additional overhead. This is because `compute_style_properties(` is called 85 times on the home page out of which 83 calls had a unique signature, so it was as good as having no cache at all + cache key generation overhead. The same -ve effect was found for the non-cached site. Overall the code resulted in a `6%` degrade.

<img width="268" alt="Screenshot 2024-04-15 at 14 44 44" src="https://github.com/WordPress/wordpress-develop/assets/30643833/ffe94bfd-f2d1-4705-b47a-c94e12bffc4b">


### **No cache**  - All score below in (ms)
Trunk Response Time (median) - 399.62
After PR - 424.81 -- Note this cost is mainly due to cache key creation code run separately 85 times

### **Meme cache with `theme_json` non persistent cache**
Trunk Response Time (median - 375.47
After PR with cache - 402.58


## Switching to transient-based caching ( [this commit ](https://github.com/WordPress/wordpress-develop/pull/6392/commits/3bba97f64b17f69692912faff6c9605221c2c598))
The way around this was to use transient function-based caching. The idea was transient would override the -ve effect of overhead due to cache key creation on non-persistent sites and it would give the benefit of a persistent cache for other sites.

Testing the above concept [this commit ](https://github.com/WordPress/wordpress-develop/pull/6392/commits/3bba97f64b17f69692912faff6c9605221c2c598) - 

### **No cache** 
Trunk - 400 
PR (transient cache) - 471.61  ( mainly because get_transient_site is costly) + cache_key code.

### **Meme cache**
Trunk - 374
PR - 376 (benefit negated by wp_json_encode and md5 )


## Improved transient caching through static caching (current state)
### **Meme cache**
Trunk - 374
PR - 356 (~ 6% improvement )

### **No cache** 
Trunk - 400 
PR (transient cache) - 384  ( ~4% improvement )




---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

kt-12 committed:
add caching

kt-12 committed:
transient implementation

kt-12 committed:
Improve caching key computation

kt-12 commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2128,9 +2144,33 @@ protected static function compute_style_properties( $styles, $settings = array()
+	 *
+	 * @return string The hash of the global settings.
+	 */
+	private function wp_get_global_settings_hash() {
```
`get_global_settings` is a part of `theme_json` group. This function matches with the lifecycle of the hashed array. 
`md5` cost increases with an increase in the size of the string. Since we were computing md5 almost every time (85 times) for this large array, we had an unnecessary overhead. Caching the hash resulted in a `6%` improvement ( initially response median was 377, after this code it was 354 ).

kt-12 commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2053,6 +2053,22 @@ protected static function compute_style_properties( $styles, $settings = array()
 
+		$can_use_cached = ! wp_is_development_mode( 'theme' );
+		if ( $can_use_cached ) {
+			$func_args_hash = md5( wp_json_encode( func_get_args() ) ); // phpcs:ignore PHPCompatibility.FunctionUse.ArgumentFunctionsReportCurrentValue.Changed
```
I did a small experiment interchanging `wp_json_encode` with `serialize`.
On average `serialize` was `15%` faster than `wp_json_encode`. There were a few random outliers in 85 runs where  `wp_json_encode` performed better. 

However, the string size produced by `serialize` was almost 2x the size of `wp_json_encode`. 

The cost of `md5` increases with an increase in string size. 
On the current line (2071) `seralize + md5` and `wp_json_encode + md5` had near similar cost, but in `wp_get_global_settings_hash`  the cost of `wp_json_encode + md5` was lower compared to the counterpart.

kt-12 committed:
improving performace for non persistance cache.

kt-12 committed:
cache related call only for non development mode

kt-12 committed:
cache clear code

kt-12 commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2128,9 +2156,73 @@ protected static function compute_style_properties( $styles, $settings = array()
+	 *
+	 * @since 6.6.0
+	 */
+	public static function clear_compute_style_properties_cache() {
```
There is a known bug here. The current code doesn't regenerate `compute_style_properties_cache` after clearing.

kt-12 committed:
change cache invalidating logic

kt-12 committed:
remove get compute style from constructor

kt-12 committed:
remove uncessary check

kt-12 commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2128,9 +2154,76 @@ protected static function compute_style_properties( $styles, $settings = array()
+	 *
+	 * @return string The hash of the global settings.
+	 */
+	private static function wp_get_global_settings_hash() {
```
To read: https://github.com/WordPress/wordpress-develop/pull/6392#discussion_r1566571741

kt-12 committed:
attach clear_compute_style_properties_cache directly to hook

joemcgill commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2053,6 +2061,24 @@ protected static function compute_style_properties( $styles, $settings = array()
+			* The global settings can include dynamic data related to typography.
+			* We need to evaluate it so that the cache is invalidated when it changes.
+			* Computing global settings hash is costly, so we cache it in a non-persistent cache.
+			*/
```
```suggestion
			/*
			 * The global settings can include dynamic data related to typography.
			 * We need to evaluate it so that the cache is invalidated when it changes.
			 * Computing global settings hash is costly, so we cache it in a non-persistent cache.
			 */
```

joemcgill commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2128,9 +2154,76 @@ protected static function compute_style_properties( $styles, $settings = array()
+			return $cache;
+		}
+
+		$hash = md5( wp_json_encode( wp_get_global_settings() ) );
```
This would create the same circular dependency issue that was raised in [this comment](https://github.com/WordPress/wordpress-develop/pull/5567#discussion_r1452068125). If the computed style properties we want to cache are related to the final computed `WP_Theme_JSON` object, then we would need to move this cache to the place where that final object is accessed. It looks like this happens in `wp_add_global_styles_for_blocks()` where the computed `WP_Theme_JSON` is returned by `WP_Theme_JSON_Resolver::get_merged_data()` and then that instance is used to get a list of block nodes and loop through all of the blocks to get the styles for each block.

Otherwise, we need to look into what theme json property from `wp_get_global_settings()` could be used as part of the cache strategy here.

joemcgill commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2128,9 +2154,76 @@ protected static function compute_style_properties( $styles, $settings = array()
+	 *
+	 * @since 6.6.0
+	 */
+	public static function clear_compute_style_properties_cache() {
```
We would probably need to call this when other `WP_Theme_JSON` related caches are cleared, in `wp_clean_theme_json_cache()`.

kt-12 replied:
I did try that but it was breaking an old test case `test_get_user_data_from_wp_global_styles_does_not_use_uncached_queries_for_logged_out_users` where the count of queries executed was taken into consideration. But yes I can move the call to `clear_compute_style_properties_cache` there.

joemcgill reviewed (changes requested):
Thanks @kt-12. I can confirm that this makes a big improvement in my profiling, so it would be great to keep some focus on this problem. The biggest issue with the current approach is the use of `wp_add_global_styles_for_blocks()` inside the `WP_Theme_JSON` class. We need to find an approach to caching this data that does't require the use of this global function or rethink the approach a bit.

kt-12 committed:
Update src/wp-includes/class-wp-theme-json.php

cs fixes

Co-authored-by: Joe McGill <801097+joemcgill@users.noreply.github.com>

kt-12 committed:
global style caching

kt-12 committed:
update old test

kt-12 committed:
remove additional function

kt-12 committed:
update function name

kt-12 committed:
cs fixes

kt-12 committed:
cache is set check

kt-12 committed:
cache is set check

kt-12 committed:
assert if it's not empty after

joemcgill commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -306,8 +306,30 @@ function wp_add_global_styles_for_blocks() {
+
+	$can_use_cached = ! wp_is_development_mode( 'theme' );
+	if ( $can_use_cached ) {
+		$settings_hash = _wp_get_global_settings_hash();
```
Since we're not reusing this hash anywhere else, we can probably keep all of the hash generation logic here, rather creating a separate private function. It would also be helpful to add an inline comment explaining why `wp_get_global_settings()` needs to be used for cache busting.

joemcgill commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -306,8 +306,30 @@ function wp_add_global_styles_for_blocks() {
+	$can_use_cached = ! wp_is_development_mode( 'theme' );
+	if ( $can_use_cached ) {
+		$settings_hash = _wp_get_global_settings_hash();
+		$cache_key     = 'wp_styles_for_blocks';
```
Why not use the `$settings_hash` as a part of this cache key instead of adding it to each style cache?

```suggestion
		$cache_key     = "wp_styles_for_blocks:$settings_hash";
```

kt-12 replied:
It was due to the delete transition. I wanted a label that was easy to delete. But, yes I can do this and add a pattern based delete in delete transition.

joemcgill replied:
Ah, that makes sense. Given that we are implementing this as a transient with an expiration time, I don't think we need a delete transition at all. If the `$settings_hash` changes, it will naturally cause a cache miss and a new transient will be created. The old transient will naturally expire and get cleaned up when WP deletes expired transients.

kt-12 replied:
That makes sense. If something changes it will change the settings too in most cases. I was wondering, Will there be cases where settings won't change but the results would; like during theme update?

joemcgill replied:
That's a good question. I think the assumption being made in your PR so far is that the combination of the hashed global settings object in combination with the hashed block metadata for each block, is enough of a fingerprint to use for cache busting. I think this is probably correct, but it occurred to me that rather than hashing each block node individually, we could also hash the full `$block_nodes` object as part of the transient key and avoid hashing during each instance of the loop, since changes to any blocks would invalidate the whole cache. I've demonstrated this and some of my other suggestions in the following diff:

```diff
diff --git a/src/wp-includes/global-styles-and-settings.php b/src/wp-includes/global-styles-and-settings.php
index 639c99b409..901dc95e68 100644
--- a/src/wp-includes/global-styles-and-settings.php
+++ b/src/wp-includes/global-styles-and-settings.php
@@ -309,23 +309,26 @@ function wp_add_global_styles_for_blocks() {
 
 	$can_use_cached = ! wp_is_development_mode( 'theme' );
 	if ( $can_use_cached ) {
+		$block_hash    = md5( wp_json_encode( $block_nodes ) );
 		$settings_hash = _wp_get_global_settings_hash();
-		$cache_key     = 'wp_styles_for_blocks';
+		$cache_key     = "wp_styles_for_blocks:$block_hash:$settings_hash";
 		$cached        = get_site_transient( $cache_key );
 		if ( ! is_array( $cached ) ) {
 			$cached = array();
 		}
 	}
 
+	$update_cache = false;
+
 	foreach ( $block_nodes as $metadata ) {
 		if ( $can_use_cached ) {
-			$metadata_hash   = md5( wp_json_encode( $metadata ) );
-			$style_cache_key = $settings_hash . '_' . $metadata_hash;
-			if ( isset( $cached[ $style_cache_key ] ) ) {
-				$block_css = $cached[ $style_cache_key ];
+			$block_name = str_replace( '/', '-', $metadata['name'] );
+			if ( isset( $cached[ $block_name ] ) ) {
+				$block_css = $cached[ $block_name ];
 			} else {
-				$block_css                  = $tree->get_styles_for_block( $metadata );
-				$cached[ $style_cache_key ] = $block_css;
+				$block_css             = $tree->get_styles_for_block( $metadata );
+				$cached[ $block_name ] = $block_css;
+				$update_cache          = true;
 			}
 		} else {
 			$block_css = $tree->get_styles_for_block( $metadata );
@@ -351,7 +354,7 @@ function wp_add_global_styles_for_blocks() {
 			if ( str_starts_with( $metadata['name'], 'core/' ) ) {
 				$block_name   = str_replace( 'core/', '', $metadata['name'] );
 				$block_handle = 'wp-block-' . $block_name;
-				if ( in_array( $block_handle, $wp_styles->queue ) ) {
+				if ( in_array( $block_handle, $wp_styles->queue, true ) ) {
 					wp_add_inline_style( $stylesheet_handle, $block_css );
 				}
 			} else {
@@ -366,7 +369,7 @@ function wp_add_global_styles_for_blocks() {
 				if ( str_starts_with( $block_name, 'core/' ) ) {
 					$block_name   = str_replace( 'core/', '', $block_name );
 					$block_handle = 'wp-block-' . $block_name;
-					if ( in_array( $block_handle, $wp_styles->queue ) ) {
+					if ( in_array( $block_handle, $wp_styles->queue, true ) ) {
 						wp_add_inline_style( $stylesheet_handle, $block_css );
 					}
 				} else {
@@ -376,7 +379,7 @@ function wp_add_global_styles_for_blocks() {
 		}
 	}
 
-	if ( $can_use_cached ) {
+	if ( $update_cache ) {
 		set_site_transient( $cache_key, $cached, HOUR_IN_SECONDS );
 	}
 }

```

kt-12 replied:
@joemcgill While this code should have worked, strangely, some `$metadata` doesn't contain a block name, causing the code to fail. 
This is one sample array I got, 
```
Array
(
    [path] => Array
        (
            [0] => styles
            [1] => blocks
            [2] => core/navigation
            [3] => elements
            [4] => link
        )

    [selector] => .wp-block-navigation a:where(:not(.wp-element-button))
)
```

I am checking to see if we can skip those blocks entirely from logic. It might make the code more performant.

joemcgill commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -306,8 +306,30 @@ function wp_add_global_styles_for_blocks() {
-		$block_css = $tree->get_styles_for_block( $metadata );
+		if ( $can_use_cached ) {
+			$metadata_hash   = md5( wp_json_encode( $metadata ) );
+			$style_cache_key = $settings_hash . '_' . $metadata_hash;
+			if ( isset( $cached[ $style_cache_key ] ) ) {
+				$block_css = $cached[ $style_cache_key ];
+			} else {
+				$block_css                  = $tree->get_styles_for_block( $metadata );
+				$cached[ $style_cache_key ] = $block_css;
+			}
```
If the settings hash is part of the transient key, it doesn't need to be used for each block node key:

```suggestion
			if ( isset( $cached[ $metadata_hash ] ) ) {
				$block_css = $cached[ $metadata_hash ];
			} else {
				$block_css                  = $tree->get_styles_for_block( $metadata );
				$cached[ $metadata_hash ] = $block_css;
			}
```

joemcgill commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -353,6 +375,10 @@ function wp_add_global_styles_for_blocks() {
 	}
+
+	if ( $can_use_cached ) {
+		set_site_transient( $cache_key, $cached, HOUR_IN_SECONDS );
```
This will cause the transient expiry time to get bumped even when the cached value isn't changing, which also results in 2 additional DB queries, so we should only update the transient if after checking whether the cached value has changed.

kt-12 replied:
Yes, you are correct. Thanks for pointing this out.

spacedmonkey replied:
Once https://github.com/WordPress/wordpress-develop/pull/6311/files is merged, that will be just one query.

joemcgill reviewed (dismissed):
I like the idea here. I've left a few suggestions for how I think this could be improved.

joemcgill committed:
Merge branch 'WordPress:trunk' into enhancement/object-cache-compute-style-properties

joemcgill commented on tests/phpunit/tests/theme/wpThemeJsonResolver.php:
```diff
@@ -703,7 +703,8 @@ public function test_get_user_data_from_wp_global_styles_does_not_use_uncached_q
 		}
 		$query_count = get_num_queries() - $query_count;
-		$this->assertSame( 0, $query_count, 'Unexpected SQL queries detected for the wp_global_style post type prior to creation.' );
+		// we have transient delete query inside `wp_clean_theme_json_cache()` which result in 3 queries.
```
Since the transients should expire naturally, we can probably avoid these extra queries. See [related comment](https://github.com/WordPress/wordpress-develop/pull/6392/files#r1608427419).

joemcgill commented on tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php:
```diff
@@ -73,6 +73,29 @@ public function test_third_party_blocks_inline_styles_get_registered_to_global_s
+			'.wp-block-my-third-party-block{background-color: hotpink;}',
+			$this->get_global_styles(),
+			'Third party block inline style should be registered after running wp_add_global_styles_for_blocks()'
+		);
```
I don't think this check is necessary for this test, since this is duplicating the `test_third_party_blocks_inline_styles_get_registered_to_global_styles_when_per_block()` method below and is unrelated to checking whether the transient was set.

joemcgill commented on tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php:
```diff
@@ -73,6 +73,29 @@ public function test_third_party_blocks_inline_styles_get_registered_to_global_s
+	/**
+	 * @ticket 59595
+	 */
+	public function test_check_cache_is_set() {
```
I think the name could more clearly describe what cache is being tested.

```suggestion
	public function test_styles_for_blocks_cache_is_set() {
```

joemcgill replied:
Handled in [d0b243c](https://github.com/WordPress/wordpress-develop/pull/6392/commits/d0b243cdd828e1ae3d03ad206918574de9fab94b).

joemcgill committed:
Update unit tests

joemcgill committed:
Fix typo

joemcgill commented:
@kt-12 I spent some time looking over the unit tests and add a couple additional tests that I think could be useful in [d0b243c](https://github.com/WordPress/wordpress-develop/pull/6392/commits/d0b243cdd828e1ae3d03ad206918574de9fab94b). I could not think of a good way of ensuring that the cache was actually used in order to avoid unnecessary calls to `get_styles_for_block()` since we're not able to modify the `WP_Theme_JSON_Resolver` class  hard coded in `wp_get_global_styles_custom_css()` nor the `WP_Theme_JSON` object that is returned by `WP_Theme_JSON_Resolver::get_merged_data()`.

kt-12 committed:
update cache related chanage

kt-12 committed:
remove repetition

kt-12 committed:
Merge branch 'trunk' into enhancement/object-cache-compute-style-properties

kt-12 committed:
update style block caching logic

kt-12 committed:
update test cases.

kt-12 committed:
revert test changes

kt-12 commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -306,8 +306,47 @@ function wp_add_global_styles_for_blocks() {
+		}
+
+		if ( $can_use_cached && $block_name ) {
+			if ( isset( $cached[ $block_name ] ) ) {
```
@joemcgill  I removed `str_replace( '\', '-'`. I felt it was an unnecessary overhead, given that we only need a valid string here, which we get from `$block_name` itself.

joemcgill commented:
This is looking good. Just doing some testing of the PR and I see a large improvement in the `wp_add_global_styles_for_blocks()` function. Some of the improvement is offset by the fact that the transient requires another DB query, but in persistent caches, those will also be avoided.

**Trunk**

<img width="1512" alt="image" src="https://github.com/WordPress/wordpress-develop/assets/801097/ff5152cd-8ec4-4359-a7c3-df6dc84ab5d3">

**This PR**
<img width="1512" alt="image" src="https://github.com/WordPress/wordpress-develop/assets/801097/34dc8ee7-4c0c-4807-8036-dfd0532cec58">

Some of the Unit test failures seem legitimate, so let's figure out those issues and this should be almost ready to go in.

kt-12 committed:
unregster bug fix

kt-12 commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -1000,6 +1000,14 @@ protected static function get_blocks_metadata() {
+				unset( static::$blocks_metadata[ $block_name ] );
+			}
+		}
+
```
We haven't been clearing unregistered blocks till now. This caused inconsistent outcomes, especially when the cache was in place.

kt-12 commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -1000,6 +1000,14 @@ protected static function get_blocks_metadata() {
 		$registry = WP_Block_Type_Registry::get_instance();
 		$blocks   = $registry->get_all_registered();
 
+		// Unset old blocks from static variable.
```
We haven't been clearing unregistered blocks till now. This caused inconsistent outcomes, for `WP_Theme_JSON::get_styles_block_nodes()`

joemcgill replied:
Good find. I'm really unsure if having stale metadata stored in the `WP_Theme_JSON` class is really a problem, as it seems like most of the places that make use of this data are doing so for validation purposes when comparing against the active `WP_Theme_JSON` data.

I am curious if you still see the behavior that you noticed if the changes from #6696 are applied.

If not removing this metadata leads to a bug, it would be best to document the specifics of the bug with steps to reproduce and propose this change outside the scope of this PR.

joemcgill committed:
Clear theme json data between theme unit tests

joemcgill commented on tests/phpunit/tests/theme/wpAddGlobalStylesForBlocks.php:
```diff
@@ -20,6 +20,7 @@ class Tests_Theme_WpAddGlobalStylesForBlocks extends WP_Theme_UnitTestCase {
 
 	public function set_up() {
 		parent::set_up();
+		switch_theme( 'default' );
```
This and the similar change in the `tear_down()` method are not necessary if #6696 is committed. The reason this fixes the cross-pollution you're seeing is that `wp_clean_theme_json_cache()` gets called on the `switch_theme` action and I think it's better to fix this in the main `WP_Theme_UnitTestCase` method instead of in each test class that extends that method.

joemcgill committed:
Revert "unregster bug fix"

This reverts commit 48a310c65c0b17dcda9459a2fcde97c9bee687f4.

joemcgill committed:
Merge branch 'trunk' into update/59595-fix-tests

joemcgill committed:
Don't cache nodes without names

joemcgill commented on src/wp-includes/global-styles-and-settings.php:
```diff
@@ -336,22 +375,10 @@ function wp_add_global_styles_for_blocks() {
 		}
+	}
 
-		// The likes of block element styles from theme.json do not have  $metadata['name'] set.
```
After doing some more debugging. I believe this change, along with trying to resolve the block name cache key using the `$metadata['path']` property is what is causing the remaining test failures here. In this particular case, there are styles included in the theme.json data for the `my/third-party-block` block followed by custom styles targeting the `cite` element inside that block as separate block nodes. When the path is used to resolve the block name used as the key, the first node gets added to the cache, causing the second one to be skipped.

I'll open a separate PR with an example of how to fix this, but it means that only nodes with name keys would get cached. In retrospect, maybe your original idea of using a hash of each node as the key was a better solution as long as it doesn't cause a large performance penalty.

joemcgill committed:
Merge branch 'trunk' into enhancement/object-cache-compute-style-properties

joemcgill committed:
Merge branch 'enhancement/object-cache-compute-style-properties' into update/59595-fix-tests

kt-12 committed:
remove test case

joemcgill commented on tests/phpunit/tests/theme/base.php:
```diff
@@ -45,6 +45,7 @@ public function tear_down() {
 
 		wp_clean_themes_cache();
 		wp_clean_theme_json_cache();
+		wp_clean_theme_json_cache();
```
This is now redundant and can be removed.

kt-12 replied:
Fixed now.

kt-12 committed:
cache node with no name

kt-12 committed:
test cases fixes

joemcgill committed:
Fix enqueueing inline styles for nodes without block names.

This reinstates the code block that handled proper inlining of styles that are added through theme.json files and updates the way block nodes are being cached so that we use the block name as the cache key when available, or a hash of nodes that don't have a block name because they've been added through theme.json files.

joemcgill commented:
@kt-12 I went ahead and resolved conflicts and updated this branch with the fixes proposed in [my other PR](https://github.com/WordPress/wordpress-develop/pull/6701) so will close that now. Looks like this has fixed the test failures we were seeing on this branch. Let's do some final checks, but I think this is about ready to commit.

joemcgill commented:
With the latest updates this is still profiling really well and showing a large improvement for `wp_add_global_styles_for_blocks()`.

**Trunk**
<img width="1289" alt="image" src="https://github.com/WordPress/wordpress-develop/assets/801097/14bf909a-6a20-4d1a-9dcd-f02ddd4dc9fb">

**PR**
<img width="1287" alt="image" src="https://github.com/WordPress/wordpress-develop/assets/801097/ea22be7b-0774-49ce-b461-19be6f287457">

That said, those improvements are not currently reflected in the benchmarks performed by [the Performance Workflow](https://github.com/WordPress/wordpress-develop/actions/runs/9356664409). I suspect it's because the extra query being performed to get the transient is offsetting the savings introduced in the function, but that's just a guess. Given the size of the potential improvement, it still may be worth committing so we can continue to do further testing during the beta period.

joemcgill committed:
Inline docs improvements

joemcgill commented:
Merged in [58334](https://core.trac.wordpress.org/changeset/58334).
